pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+		// #[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))] TODO: properly configure weights
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			curve: Curve,
+			currencies: BoundedVec<TokenMeta<FungiblesBalanceOf<T>, FungiblesAssetIdOf<T>>, T::MaxCurrencies>,
+			frozen: bool,
+			// currency_admin: Option<AccountIdLookupOf<T>> TODO: use this to set currency admin
+		) -> DispatchResultWithPostInfo {
+			// ensure origin is PoolCreateOrigin
+			let who = T::PoolCreateOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				(2..=T::MaxCurrencies::get().saturated_into()).contains(&currencies.len()),
+				Error::<T>::CurrenciesOutOfBounds
+			);
+
+			let currency_ids = BoundedVec::truncate_from(currencies.iter().map(|c| c.id.clone()).collect());
+
+			let pool_id = T::PoolId::from(currency_ids.blake2_256());
+
+			T::DepositCurrency::hold(
+				&pool_id.clone().into(), // TODO: just assumed that you can use a pool id as hold reason, not sure that's true though
+				&who,
+				T::DepositPerCurrency::get()
+					.saturating_mul(currencies.len().saturated_into())
+					.saturated_into(),
+			)?;
+
+			for entry in currencies {
+				let asset_id = entry.id.clone();
+
+				// create new assset class; fail if it already exists
+				T::Fungibles::create(asset_id.clone(), pool_id.clone().into(), false, entry.min_balance)?;
+
+				// set metadata for new asset class
+				T::Fungibles::set(
+					asset_id,
+					&pool_id.clone().into(),
+					entry.name.clone(),
+					entry.symbol.clone(),
+					entry.decimals,
+				)?;
+
+				// TODO: use fungibles::roles::ResetTeam to update currency admin
+			}
+
+			<Pools<T>>::set(
+				pool_id,
+				Some(PoolDetails::new(who.clone(), curve, currency_ids, !frozen)),
+			);
+
+			// Emit an event.
+			Self::deposit_event(Event::PoolCreated(who));
+			// Return a successful DispatchResultWithPostInfo
+			Ok(().into())
+		}
+
+		#[pallet::call_index(1)]
+		pub fn mint_into(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			currency_idx: u32,
+			amount_to_mint: FungiblesBalanceOf<T>,
+			max_cost: CollateralCurrencyBalanceOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let signer = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			let pool_details = <Pools<T>>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			let mint_enabled = match pool_details.state {
+				// if mint is locked, then operation is priviledged
+				PoolStatus::Frozen(locks) => locks.allow_mint || signer == pool_details.creator,
+				PoolStatus::Active => true,
+				_ => false,
+			};
+			ensure!(mint_enabled, Error::<T>::Locked);
+
+			let currency_idx_usize: usize = currency_idx.checked_into().ok_or(Error::<T>::IndexOutOfBounds)?;
+
+			// get id of the currency we want to mint
+			// this also serves as a validation of the currency_idx parameter
+			let mint_currency_id = pool_details
+				.bonded_currencies
+				.get(currency_idx_usize)
```

Comment:
I guess you can simply use the currency_idx with the saturation converting.

Reply:
we use it twice so why convert it twice?

Reply:
Oh, okay, did not see that